plugins/org.eclipse.n4js.xpect.ui/src/org/eclipse/n4js/xpect/ui/refactoring/RenameRefactoringXpectMethod.java

```diff
@@ -0,0 +1,190 @@
+import org.eclipse.xtext.ui.refactoring.ui.IRenameElementContext;
+import org.eclipse.xtext.ui.refactoring.ui.IRenameSupport;
+
+import com.google.inject.Inject;
+
+/**
+ * Provides XPEXT test methods for rename refactoring
+ */
+@XpectImport({ N4JSOffsetAdapter.class, XpEnvironmentData.class, VarDef.class, Config.class,
+		ValidationTestModuleSetup.class })
+@SuppressWarnings("restriction")
+public class RenameRefactoringXpectMethod {
+
+	@Inject
+	private IRenameSupport.Factory renameSupportFactory;
+
+	@Inject
+	private EObjectAtOffsetHelper offsetHelper;
+
+	@Inject
+	private IRenameContextFactory renameContextFactory;
+
+	static {
+		EclipseGracefulUIShutdownEnabler.enableOnce();
+	}
+
+	/**
+	 * Rename refactoring Xpect method
+	 */
+	@ParameterParser(syntax = "('at' arg2=OFFSET 'to' arg3=STRING) ('resource' arg4=STRING)?") // arg1=OFFSET makes the
+																								// 'offset' parameter
+																								// contain
+	// the right offset value
+	@Xpect
+	@ConsumedIssues({ Severity.INFO, Severity.ERROR, Severity.WARNING })
+	public void renameRefactoring(
+			@StringDiffExpectation(whitespaceSensitive = false) IStringDiffExpectation expectation, // arg0
+			@ThisResource XtextResource resource, // arg1
+			IEObjectCoveringRegion offset, // arg2
+			String newName, // arg3
+			String specifiedResourcePath, // arg4
+			@N4JSCommaSeparatedValuesExpectation IN4JSCommaSeparatedValuesExpectation expectedResult)
+			throws Exception {
+
+		EObject context = offset.getEObject();
+		EObject selectedElement = offsetHelper.resolveElementAt((XtextResource) context.eResource(),
+				offset.getOffset());
+
+		// GH-1002: Ask Oliver
+		// Special handling for FormalParameter
+		// IdentifierRef refers to FormalParameter
+		EObject selectedTypeElement = (selectedElement instanceof FormalParameter) ? null
+				: N4JSLanguageUtils.getDefinedTypeModelElement(selectedElement);
+		selectedTypeElement = selectedTypeElement == null ? selectedElement : selectedTypeElement;
+
+		URI targetResourceUri = context.eResource().getURI();
+		// XtextResource resource = (XtextResource) context.eResource();
+
+		N4JSGlobals.myGlobalResourceSet = context.eResource().getResourceSet();
+
+		Optional<XtextEditor> editorOp = EditorsUtil.openXtextEditor(targetResourceUri,
+				N4JSActivator.ORG_ECLIPSE_N4JS_N4JS);
+		XtextEditor editor = editorOp.get();
+
+		final ITextSelection selection = (ITextSelection) editor.getSelectionProvider().getSelection();
+
+		IRenameElementContext renameElementContext = renameContextFactory
+				.createRenameElementContext(
+						selectedTypeElement, editor, selection, resource);
+
+		IRenameSupport renameSupport = renameSupportFactory.create(renameElementContext, newName);
+
+		// HACK, use reflection to obtain the private field 'renameRefactoring' since we need it to verify the
+		// conditions
+		Field field = renameSupport.getClass().getDeclaredField("renameRefactoring");
```

Comment:
does it throw exceptions in case of a missing field?

Reply:
I think an exception is thrown by JVM but I am not sure.